Sources/_StringProcessing/Engine/Processor.swift

```diff
@@ -720,7 +716,7 @@ extension String {
     }
 
     let idx = unicodeScalars.index(after: pos)
-    guard idx <= end else { return nil }
+    assert(idx <= end, "Input is a substring with a sub-scalar endIndex.")
```

Comment:
Can we also put the assertion near the top of Processor proper? We'll want to rely on it throughout the engine